src/board/system76/common/flash/main.c

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (C) 2020 Evan Lojewski
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ */
+
+#include <stdint.h>
+#include <board/flash.h>
+
+// EC indirect flash access
+volatile uint8_t __xdata __at(0x103B) ECINDAR0;
+volatile uint8_t __xdata __at(0x103C) ECINDAR1;
+volatile uint8_t __xdata __at(0x103D) ECINDAR2;
+volatile uint8_t __xdata __at(0x103E) ECINDAR3;
+volatile uint8_t __xdata __at(0x103F) ECINDDR;
+
+#define SPI_DEVICE                  (0x70)
+#define SPI_FOLLOW_MODE             (0x0F)
+#define SPI_CHIP_SELECT             (0xFD)
+#define SPI_CHIP_DESELECT           (0xFE)
+
+#define SPI_READ_STATUS_COMMAND     (0x05)
+#define SPI_READ_COMMAND            (0x0B)
+#define SPI_WRITE_COMMAND           (0x02)
+#define SPI_WRITE_ENABLE_COMMAND    (0x06)
+#define SPI_READ_STATUS_COMMAND     (0x05)
+
+#define SPI_ERASE_SECTOR_COMMAND    (0xD7)
+
+#define SPI_STATUS_WIP              (0x01)
+
+void flash_write_enable(void);
+void flash_wait(void);
+void flash_enter_follow_mode(void);
+void flash_exit_follow_mode(void);
```

Comment:
Reorder functions before `flash_entry()` to remove the delcarations

Reply:
So, sdcc doesn't support compiling functions within the same fine in separate sections like gcc/clang based compilers do (text.function_name vs text).
As a result, the function order in the source file is the order that they will appear in the output file. Since the main code assumes flash_entry is at the beginning, I can't just reorder functions in this file.

I can leave it as is, or change this to be something like the following:
(1)
- Split main.c into two files. One will be for the main entry point. This will be placed in the GSINIT0 area to ensure that the file is *always* first in the resulting binary. Note that flash_entry *must* be first in this file. 
- Add a seperate flash_support.c (and flash_support.h file). This is compiled with the default code section. Function prototypes area moved to the .h file, as well as any common command ids, defines, etc.
(2)
Alternatively, I can add a separate crt file in GSINIT0 that jumps to flash_entry.
(3)
Add build system support code to extract the flash function addresses from the flash binary, and pass those to the main firmware image. This would remove the need for flash_entry entirely (each command can be a separate function instead), however it complected the build system a bit. This would also have slightly lower overhead as well.
(4)
Finally, I can leave it as is, just with additional comments clearly stating why flash_entry must be the first function in the file and while main.c must be the first (or only) source file. The main reason I chose this path is because it's easiest to implement.

Let me know your preference.

Reply:
Go with 4.

Reply:
Ok, comments added / expanded on. Let me know if you want any other changes related to this.